core/src/main/java/com/graphhopper/routing/AlternativeRouteCH.java

```diff
@@ -0,0 +1,253 @@
+/*
+ *  Licensed to GraphHopper GmbH under one or more contributor
+ *  license agreements. See the NOTICE file distributed with this work for
+ *  additional information regarding copyright ownership.
+ *
+ *  GraphHopper GmbH licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except in
+ *  compliance with the License. You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package com.graphhopper.routing;
+
+import com.carrotsearch.hppc.IntIndexedContainer;
+import com.carrotsearch.hppc.predicates.IntObjectPredicate;
+import com.graphhopper.routing.ch.Path4CH;
+import com.graphhopper.routing.ch.PreparationWeighting;
+import com.graphhopper.routing.weighting.Weighting;
+import com.graphhopper.storage.Graph;
+import com.graphhopper.storage.SPTEntry;
+import com.graphhopper.util.EdgeIteratorState;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.TreeSet;
+
+/**
+ *
+ * Minimum number-of-moving-parts implementation of alternative route search with
+ * contraction hierarchies.
+ *
+ * "Alternative Routes in Road Networks" (Abraham et al.)
+ *
+ * @author michaz
+ */
+public class AlternativeRouteCH extends DijkstraBidirectionCHNoSOD {
+
+    private double maxWeightFactor = 1.4;
+    private double maxShareFactor = 0.6;
+    public static final double T = 10000.0;
+
+    public AlternativeRouteCH(Graph graph, Weighting weighting) {
+        super(graph, weighting);
+    }
+
+    List<AlternativeInfo> calcAlternatives(final int s, final int t) {
+        // First, do a regular bidirectional route search
+        createAndInitPath();
+        init(s, 0, t, 0);
+        runAlgo();
+        extractPath();
+
+        final TreeSet<AlternativeInfo> alternatives = new TreeSet<>(new Comparator<AlternativeInfo>() {
+            @Override
+            public int compare(AlternativeInfo o1, AlternativeInfo o2) {
+                return Double.compare(o1.path.getWeight(), o2.path.getWeight());
+            }
+        });
+        alternatives.add(new AlternativeInfo(bestPath, 0));
+
+        final double maxWeight = maxWeightFactor * bestPath.getWeight();
+        bestWeightMapFrom.forEach(new IntObjectPredicate<SPTEntry>() {
```

Comment:
You basically to a normal point-to-point query and then you consider all other nodes where the fwd/bwd search trees overlap by the time the point-to-point query has finished ? And other (possible longer) routes where the search trees do not overlap already are discarded ?

Reply:
Nitpick: They don't overlap, they only touch. There are no plateaus in CH.

You're right, I'm sure we could find more candidates by searching a bit longer, but (as you say) what is the right bit? :) To me, it looks like we're getting just enough candidates to filter them down carefully, without missing much.

But this is totally open for experimentation.